File: BinaryWatch5.py
class Solution:
    def readBinaryWatch(self, num):
        """
        :type num: int
        :rtype: List[str]
        """
        def getTime(bytes):
            hour = 0
            for i in range(4):
                if bytes[i]:
                    hour |= (1 << i)
            minute = 0
            for i in range(4, 10):
                if bytes[i]:
                    minute |= (1 << (i - 4))
            if hour <12 and minute < 60:
                return '%d:%02d' % (hour, minute)
            else:
                return None

        def helper(count, i):
            if count == 0:
                t = getTime(bytes)
                if t:
                    result.append(t)
                return
            if i >= 10:
                return
            bytes[i] = 0
            helper(count, i + 1)
            bytes[i] = 1
            helper(count - 1, i + 1)
            bytes[i] = 0

        bytes = [0] * 10
        result = []
        helper(num, 0)
        return result


if __name__ == '__main__':
    num = int(input())
    solution = Solution()
    print(solution.readBinaryWatch(num))

File: BinaryWatch5.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<string> readBinaryWatch(int num) {
    	
    	vector<string> result;
    	vector<int> bytes(10, 0);
    	
    	helper(bytes, num, 0, result);
		
		return result;
    }
    
    void helper(vector<int>& bytes, int num, int i, vector<string>& result) {
    	if(num == 0) {
    		string t = getTime(bytes);
    		if(t.size() > 0) result.push_back(t);
    		return;
		}
		
		if(i >= 10) return;
		
		bytes[i] = 0;
		helper(bytes, num, i + 1, result);
		
		bytes[i] = 1;
		helper(bytes, num - 1, i + 1, result);
		bytes[i] = 0;
	}
	
	string getTime(vector<int> bytes) {
		int hour = 0;
		for(int i = 0; i < 4; i++) {
			if(bytes[i]) hour += 1 << i;
		}
		
		int minute = 0;
		for(int i = 4; i < bytes.size(); i++)
			if(bytes[i]) minute += 1 << (i - 4);
		
		if(hour < 12 && minute < 60)
			return to_string(hour) + (minute >= 10 ? ":" : ":0") + to_string(minute);
		else return "";
	}
    
};

int main() {
	
	int num;
	cin>>num;
	
	Solution *solution = new Solution();
	vector<string> result = solution->readBinaryWatch(num);
	
	for(int i = 0; i < result.size(); i++) cout<<result[i]<<"  ";
	cout<<endl;
	
	return 0;
	
}
